### ISSUE TITLE
NOD | Implement callback reporting url for NOD email notifications

### ISSUE THREAD

**Issue by dfong-adh** [department-of-veterans-affairs/va.gov-team#74832]
## Value Statement
**_As a_** Veteran who may have been unable to file an NOD after being unable to access VA.gov 
**_I want to_** get notified about next steps I can take if my email bounces
**_So that_** I understand how to file if my filing deadline is nearing or has passed. 

---
## Background Context
Spike ticket: #73397 
We want to set up a callback endpoint to be used by VA Notify so we can generate a report for tracking Veteran emails that return a hard bounce.

## Outcome, Success Measure, KPI(S), and Tracking Link
- Add here

## Acceptance Criteria
- [x] Callback url rejects invalid requests
- [x] Callback url accepts valid requests using bearer token authentication and stores email and status in a table
- [x] Table data can be used for report generation of Veteran emails that returned a hard bounce

## Engineering Notes
Callback format:
```
"id":"6ba01111-f3ee-4a40-9d04-234asdfb6abab9c",  -- this is the notification id
"reference":null,  
"to":"",  -- this is the recipient's contact info
"status":"delivered",  -- this will specify the delivery status of the notification
"created_at":"2023-01-10T00:04:25.273410Z",  
"completed_at":"2023-01-10T00:05:33.255911Z",  
"sent_at":"2023-01-10T00:04:25.775363Z",  
"notification_type":"SMS",   -- this is the channel/type of notification
"status_reason": "", -- populated if there's additional context on the delivery status
"provider":"pinpoint"
```
> VA Notify uses bearer tokens for authenticating with services when making a callback. The bearer token can be any form that the called service endpoint accepts. These should be long-lived, but not permanent, bearer tokens.
> Each service will have their own policy for updating a long-lived bearer token. Contact VA Notify if an updated bearer token is required to meet security policies.
> We recommend not hardcoding acceptance of only a single bearer token, so bearer token updates are not disruptive.

## Open questions
- Is there a specific callback url format we need to use?
   - Endpoint URL wil be `v1/nod_callbacks`
- How often will we need to change bearer tokens?
   - We plan to only use this endpoint for a short amount of time and can immediately remove the endpoint once the remediation process is complete.
- Should we create a table for the bearer tokens or store them in configuration files?
   - Configuration files for simplicity unless otherwise specified by Readiness Review process

## Tasks
- [x] Implement callback endpoint using bearer token for authentication
- [x] Store payload for failure notifications that will not be retried in a table
- [x] Provide callback endpoint url and bearer token to VA Notify team

## Definition of Done

- [ ] Meets acceptance criteria
- [ ] Passed E2E testing (90% coverage)
- [ ] Passed unit testing (90% coverage)
- [ ] Passed integration testing (if applicable)
- [ ] Code reviewed (internal)
- [ ] Submitted to staging
- [ ] Team approved production verification process
- [ ] Reviewed and approved by product and/or design
- [ ] Update Steve A. on this work.

**Comment by saderagsdale** [comment]
Yesterday's note from @dfong-adh:


> _I'm currently working on the bearer token authentication for the endpoint. I also met with Platform Security today and they mentioned that we will need a VFS review for the callback url. Eugene said it may be the same as a Readiness Review (like https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/issues/628) so I will be following up with Platform Security to confirm if this is the correct ticket template and the correct process._

**Comment by dfong-adh** [comment]
Reached out directly to Gary Fallon and confirmed that the correct ticket template is the [Readiness Review template](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/issues/new?assignees=kaipyroami&labels=security-review%2Cplatform-security%2Cplatform-security-review&projects=&template=privacy-and-security-review.md&title=Readiness+Review+%5BTeam-Name%2C+Feature-Name%5D). Will work with Eugene regarding the Readiness Review ticket.

**Comment by saderagsdale** [comment]
Waiting to hear back from [Platform Security](https://dsva.slack.com/archives/C01CJV0L9PS/p1706908022392009).

**Comment by saderagsdale** [comment]
@dfong-adh How's it going with the Readiness Review ticket and next steps?

**Comment by dfong-adh** [comment]
@saderagsdale I've added some details to the Readiness Review ticket - will reach out to @data-doge regarding the items I'm unsure of. Currently working on the backend code for the callback url and will work with @data-doge to get that deployed to staging once we have something that can log requests.

**Comment by HeatherWidmont** [comment]
Endpoint is working locally & will work with @data-doge to get it deployed to staging

**Comment by dfong-adh** [comment]
PR for updating staging environment with new secret: https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/2668/files

**Comment by saderagsdale** [comment]
@dfong-adh @data-doge Hey both. The VA notify team let me know that we'll have to figure out a way to use the callback URL to pull the emails that return a hard bounce and drop it into a spreadsheet. I think that should be as simple as checking what the endpoint has returned and placing it into a Sharepoint spreadsheet, but let me know if I'm oversimplifying. Their explanation below:

> VA Notify would be pushing updates to your URL.
> 
> Workflow Overview:
> 
> The system you use to trigger the email request will receive a notification ID at that time VA Notify accepts the request.
> VA Notify then attempts to look up the Veteran and deliver the notification
> Delivery statuses will be sent back to the URL your team provides in realtime
> Your team can store that data and determine the best way to provide the report you're looking for.

**Comment by dfong-adh** [comment]
https://github.com/department-of-veterans-affairs/vets-api/pull/15436 has been merged and is awaiting deployment to staging

Currently working on another PR to add a migration for a new table and backend logic to encrypt and store the VANotify payloads in the new table.

**Comment by HeatherWidmont** [comment]
Last PR should be ready for review today. 

@data-doge working with VA notify to test in staging

**Comment by data-doge** [comment]
Created a follow up ticket here: https://github.com/department-of-veterans-affairs/va.gov-team/issues/76113

**Comment by data-doge** [comment]
Derek drafted a final PR for this ticket, and I left a few comments. Ticket should be done very soon, at which point we can validate in staging.

**Comment by HeatherWidmont** [comment]
@dfong-adh @data-doge what is the estimate for this work?

**Comment by saderagsdale** [comment]
Pull me in when it's time to look in staging!

**Comment by dfong-adh** [comment]
Final PR has been merged and deployed to staging.
